src/main/java/org/broadinstitute/dsm/jobs/TestBostonUPSTrackingJob.java

```diff
@@ -24,124 +21,186 @@
 import java.sql.PreparedStatement;
 import java.sql.ResultSet;
-import java.sql.SQLException;
+import java.sql.Statement;
 import java.time.Instant;
-import java.util.ArrayList;
 import java.util.Arrays;
-import java.util.Base64;
 import java.util.List;
 
 
-public class TestBostonUPSTrackingJob implements BackgroundFunction<PubsubMessage> {
+public class TestBostonUPSTrackingJob implements BackgroundFunction<UPSKit[]> {
 
     private String STUDY_MANAGER_SCHEMA = System.getenv("STUDY_MANAGER_SCHEMA") + ".";
     private String STUDY_SERVER_SCHEMA = System.getenv("STUDY_SERVER_SCHEMA") + ".";
 
     private final Logger logger = LoggerFactory.getLogger(TestBostonUPSTrackingJob.class);
 
-    private String SQL_AVOID_DELIVERED = "and (tracking_to_id is not null or tracking_return_id is not null ) and (kit_shipping_history is null or kit_shipping_history not like \"" + UPSStatus.DELIVERED_TYPE + " %\" or kit_return_history is null or kit_return_history not like \"" + UPSStatus.DELIVERED_TYPE + " %\")" +
-            " order by kit.dsm_kit_request_id ASC";
-
-
     String DELIVERED = "DELIVERED";
     String RECEIVED = "RECEIVED";
 
     private String SELECT_BY_EXTERNAL_ORDER_NUMBER = "and request.external_order_number = ?";
     private Covid19OrderRegistrar orderRegistrar;
+    String endpoint;
+    String username;
+    String password;
+    String accessKey;
+    Connection conn = null;
 
     @Override
-    public void accept(PubsubMessage message, Context context) throws Exception {
-        if (message.data == null) {
+    public void accept(UPSKit[] kitsToLookFor, Context context) throws Exception {
+        if (kitsToLookFor == null || kitsToLookFor.length == 0) {
             logger.info("No message provided");
             return;
         }
-
-
-        logger.info(message.data + " *");
         Config cfg = CFUtil.loadConfig();
-
         String dbUrl = cfg.getString("dsmDBUrl");
-
-        PoolingDataSource<PoolableConnection> dataSource = CFUtil.createDataSource(20, dbUrl);
-
-
+        endpoint = cfg.getString("ups.url");
+        username = cfg.getString("ups.username");
+        password = cfg.getString("ups.password");
+        accessKey = cfg.getString("ups.accesskey");
+        PoolingDataSource<PoolableConnection> dataSource = CFUtil.createDataSource(1, dbUrl);
         logger.info("Starting the UPS lookup job");
 
+        try {
+            conn = dataSource.getConnection();
+            Arrays.stream(kitsToLookFor).forEach(
+                    kit -> {
+                        logger.info("Checking possible actions for kit " + kit.getDsmKitRequestId());
+                        if (StringUtils.isNotBlank(kit.getUpsPackage().getUpsShipmentId())) {
+                            getUPSUpdate(conn, kit, cfg);
+                        }
+                        else {
+                            insertShipmentAndPackageForNewKit(conn, kit, cfg);// for a new kit we first need to insert the UPSShipment
```

Comment:
Given a potentially large volume of kits here and the flakiness of the UPS endpoint, I think it'd be be better to do the db updates for each kit in a separate connection/transaction.  In this way we'll commit updates as we make them and minimize the chance that one UPS query explosion will rollback all updates for all kits in the batch.